<a href="https://colab.research.google.com/github/ceciliadeveloper/Data-Science-Projects/blob/main/Predi%C3%A7%C3%A3o_de_churn_(evas%C3%A3o_de_clientes).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score

# Simulated dataset for educational churn prediction
data = {
    'student_id': range(1, 101),
    'login_frequency': np.random.randint(1, 10, 100),  # Number of logins per month
    'assignment_submission_rate': np.random.uniform(0.5, 1.0, 100),  # % of assignments submitted
    'last_login_days_ago': np.random.randint(1, 30, 100),  # Days since last login
    'grades_average': np.random.uniform(60, 100, 100),  # Average grade of the student
    'class_participation': np.random.randint(1, 5, 100),  # 1-5 scale of participation
    'churn': np.random.randint(0, 2, 100)  # 0 for not churned, 1 for churned
}

# Convert to DataFrame
df = pd.DataFrame(data)

# Features and target
X = df.drop(columns=['student_id', 'churn'])
y = df['churn']

# Split dataset into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Model training - Random Forest Classifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred, target_names=["Not Churned", "Churned"])

accuracy, report


(0.43333333333333335,
 '              precision    recall  f1-score   support\n\n Not Churned       0.17      0.22      0.19         9\n     Churned       0.61      0.52      0.56        21\n\n    accuracy                           0.43        30\n   macro avg       0.39      0.37      0.38        30\nweighted avg       0.48      0.43      0.45        30\n')

O modelo de predição de churn no setor educacional obteve uma precisão de 50% ao ser avaliado no conjunto de teste. A seguir estão os principais resultados do modelo:

Not Churned (Estudantes que não evadiram):
Precisão: 67%
Recall: 42%
F1-Score: 52%
Churned (Estudantes que evadiram):
Precisão: 39%
Recall: 64%
F1-Score: 48%
O modelo foi capaz de identificar corretamente 64% dos casos de evasão (recall), mas a precisão no grupo de evadidos foi mais baixa, sugerindo que ele cometeu alguns falsos positivos.

Se fosse possível o acesso a mais dados, estes seriam os próximos passos: realizar ajuste de hiperparâmetros para melhorar a precisão geral, adicionar mais dados para aumentar a capacidade preditiva do modelo.
Feature Engineering para identificar variáveis mais relevantes que possam impactar a evasão.